### PR TITLE
Reworking EventListener v.2

### PR DESCRIPTION
**Pull request by RedstoneFuture**
**Outsourcing of methods: "isInBetween" (for the arena), spawnMissile and spawnFireball (as player); tested Ingame**

**Different small Fixes for last Changes (new Command System, Missile Name request, Keep Inventory option); tested Ingame**

**Reworked (Refactoring, Improvement the checks, Continuity check) and tested event listener Ingame**

- InventoryClickEvent EL GL PL[deleted]
- EntityExplodeEvent GL
- BlockPhysicsEvent GL
- PlayerInteractEvent GL LL
- ProjectileLaunchEvent GL
- EntityDamageByEntityEvent GL

---

EL: EndListener
GL: GameListener
LL: LobbyListener
PL: PlayerListener

### CODE REVIEW

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/PlayerEquipmentRandomizer.java [hunk 1]
```diff
@@ -68,11 +68,35 @@ public void tick() {
         }
     }
 
+    /**
+     * This method reset the countdown for the player equipment
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/PlayerEquipmentRandomizer.java]: Grammar: `reset` -> `resets`

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameBoundListener.java [hunk 2]
```diff
@@ -41,4 +45,30 @@ public boolean isInGameWorld(Location location) {
...
+    /**
+     * This method gets the interaction protection variable for a player.
+     *
+     * @param player (Player) the target player
+     */
+    protected boolean isInteractDelay(Player player) {
+        MWPlayer mwPlayer = getGame().getPlayer(player);
+        if (mwPlayer == null) return false;
+
+        return getGame().getPlayer(player).isPlayerInteractEventCancel();
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameBoundListener.java]:
```suggestion
        return mwPlayer.isPlayerInteractEventCancel();
```

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameBoundListener.java [hunk 3]
```diff
@@ -41,4 +45,30 @@ public boolean isInGameWorld(Location location) {
...
+     * a short time.
+     *
+     * @param player (Player) the target player
+     */
+    protected void setInteractDelay(Player player) {
+        MWPlayer mwPlayer = getGame().getPlayer(player);
+        if (mwPlayer == null) return;
+
+        getGame().getPlayer(player).setPlayerInteractEventCancel(true);
+        Bukkit.getScheduler().runTaskLater(MissileWars.getInstance(), () -> getGame().getPlayer(player).setPlayerInteractEventCancel(false), 20);
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameBoundListener.java]:
```suggestion
        mwPlayer.setPlayerInteractEventCancel(true);
        Bukkit.getScheduler().runTaskLater(MissileWars.getInstance(), () -> mwPlayer.setPlayerInteractEventCancel(false), 20);
```

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java [hunk 4]
```diff
@@ -90,87 +89,59 @@ public void onMove(PlayerMoveEvent event) {
     public void onExplode(EntityExplodeEvent event) {
         if (!isInGameWorld(event.getLocation())) return;
 
-        Game game = getGame();
+        if (event.getEntity().getType() != EntityType.FIREBALL) return;
+        if (!getGame().getArena().getFireballConfiguration().isDestroysPortal()) return;
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java]: This if clause is reversed
```suggestion
        if (getGame().getArena().getFireballConfiguration().isDestroysPortal()) return;
```

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java [hunk 5]
```diff
@@ -212,37 +183,48 @@ public void onJoin(PlayerArenaJoinEvent event) {
     public void onThrow(ProjectileLaunchEvent event) {
         if (!isInGameWorld(event.getEntity().getLocation())) return;
 
-        Game game = getGame();
         if (event.getEntity() instanceof Snowball) {
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java]: Maybe also use guard clause here

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java [hunk 6]
```diff
@@ -257,6 +239,7 @@ public void onRespawn(PlayerRespawnEvent event) {
             event.setRespawnLocation(team.getSpawn());
             getGame().sendGameItems(player, true);
             getGame().setPlayerAttributes(player);
+            getGame().getPlayer(player).getRandomGameEquipment().resetPlayerInterval();
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java:243]: Maybe make this configurable? Some people maybe want to configure their game to be faster and have the time not be reset upon death.

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java [hunk 7]
```diff
@@ -212,37 +183,48 @@ public void onJoin(PlayerArenaJoinEvent event) {
...
+
+            if (!(projectile.getShooter() instanceof Player)) return;
+
+            shooter = (Player) projectile.getShooter();
+
+        } else if (event.getDamager() instanceof Player) {
+
+            shooter = (Player) event.getDamager();
+
+        } else return;
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java]: Please use here also brackets. If one part of the if clause uses brackets, then the other parts should too.
```suggestion
        } else {
          return;
        }
```

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/LobbyListener.java [hunk 8]
```diff
@@ -55,34 +55,44 @@ public LobbyListener(Game game) {
...
+            // team switch:
+            if (!player.hasPermission("mw.change")) return;
 
-            if (!p.hasPermission("mw.change")) return;
+            String displayName = event.getItem().getItemMeta().getDisplayName();
 
+            // same team:
+            if (displayName.equals(getGame().getPlayer(player).getTeam().getFullname())) return;
+
+            // time to late:
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/LobbyListener.java]: Language: either time too late or too late for team change

**Review by daniel-naegele — CHANGES_REQUESTED** [review]

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java (hunk 4 above)
Comment (RedstoneFuture) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java]: If the "destroy_portal" is `false` RETURN and don't delete the portals in the next step. Make sense for me.

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java (hunk 6 above)
Comment (RedstoneFuture) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java:243]: Currently, it's the start interval. If you down't reset the interval and die while the interval is at 0 (and you lose your items at this time), you have longer to wait as wished.

```
equipment_interval:
  start: 10
  intervals_by_team_amount:
    "1": 15
    "2": 20
    "4": 25
````

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java (hunk 4 above)
Comment (daniel-naegele) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java]: The following code will remove the portals from the list of effected blocks. So with your code if `destroys_portal` is true, then the code will progress and remove the portal blocks from the list of effected blocks and the portal won't be destroyed.

Comment (RedstoneFuture) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java]: Oh, right. I misunderstood the target behind the `blockList().removeIf` action.

**Review by daniel-naegele — APPROVED** [review]

**Pull request merged**